tests/test_bids_file2.m

```diff
@@ -70,11 +72,175 @@ function test_change()
+
+end
+
+function test_bids_file_derivatives()
+
+  % GIVEN
+  filename = 'wuasub-01_ses-test_task-faceRecognition_run-02_bold.nii';
+  file = bids.File2(filename, 'use_schema', false);
+  % WHEN
+  file.prefix = '';
+  file.entities.desc = 'preproc';
+  % THEN
+  assertEqual(file.filename, 'sub-01_ses-test_task-faceRecognition_run-02_desc-preproc_bold.nii');
+
+end
+
+%% SCHEMA
+
+function test_bids_file_parsing_filename_schema_based()
+
+  % GIVEN
+  filename = 'sub-01_task-foo_run-1_bold.nii.gz';
+  use_schema = true;
+
+  % WHEN
+  file = bids.File2(filename, 'use_schema', use_schema);
+
+  % THEN
+  assert(~isempty(file.schema));
+  assertEqual(file.modality, 'func');
+  assertEqual(file.entity_required, {'sub', 'task'});
+  assertEqual(file.entity_order, ...
+              {'sub'
+               'ses'
+               'task'
+               'acq'
+               'ce'
+               'rec'
+               'dir'
+               'run'
+               'echo'
+               'part'});
+
+  assertEqual(file.prefix, '');
+  assertEqual(file.suffix, 'bold');
+  assertEqual(file.extension, '.nii.gz');
+  assertEqual(file.entities, struct('sub', '01', 'task', 'foo', 'run', '1'));
+  assertEqual(file.filename, 'sub-01_task-foo_run-1_bold.nii.gz');
+  assertEqual(file.json_filename, 'sub-01_task-foo_run-1_bold.json');
+  assertEqual(file.bids_path, fullfile('sub-01', 'func'));
+
+end
+
+function test_bids_file_parsing_structure_schema_based()
+  % entities given in any order, should be reordered according to schema
+
+  % GIVEN
+  filename.suffix = 'bold';
+  filename.ext = '.nii';
+  filename.entities = struct('sub', '01', ...
+                             'run', '02', ...
+                             'task', 'faceRecognition');
+
+  % WHEN
+  file = bids.File2(filename, 'use_schema', true);
+
+  % THEN
+  assertEqual(file.filename, 'sub-01_task-faceRecognition_run-02_bold.nii');
+
+end
+
+%% ERRORS
+
+function test_error_required_entity()
+  % GIVEN
+  filename.suffix = 'bold';
+  filename.ext = '.nii';
+  filename.entities = struct( ...
+                             'run', '02', ...
+                             'acq', '01');
+  % THEN
+  assertExceptionThrown(@()bids.File2(filename, 'use_schema', true, 'tolerant', false), ...
+                        'File2:requiredEntity');
+
+end
+
+function test_error_suffix_in_many_modalities()
+  % GIVEN
+  filename.suffix = 'events';
+  filename.ext = '.tsv';
+  filename.entities = struct('sub', '01', ...
+                             'task', 'faces');
+  % THEN
+  assertExceptionThrown(@()bids.File2(filename, 'use_schema', true,  'tolerant', false), ...
+                        'File2:manyModalityForsuffix');
+end
+
+function test_error_no_suffix()
+  % GIVEN
+  filename.entities = struct('sub', '01', ...
+                             'task', 'faces');
+  % THEN
+  assertExceptionThrown(@()bids.File2(filename, 'use_schema', false,  'tolerant', false), ...
```

Comment:
Do we want to throw errors when there is no suffix when not using the schema?

Reply:
by the way this test fails: did not want to finish this until we are sure about the desired behavior